pkg/kubelet/apis/pods/server.go

```diff
@@ -0,0 +1,437 @@
+func (s *PodsServer) OnPodStatusUpdated(pod *v1.Pod, status v1.PodStatus) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if storedPod, ok := s.pods[pod.UID]; ok {
+		storedPod.Status = status
+		s.broadcaster.Broadcast(PodWatchEvent{Type: watch.Modified, Pod: storedPod})
+		klog.Infof("Pod %s status updated in storage", pod.UID)
+	}
+}
+
+// Get returns a pod by UID.
+func (s *PodsServer) Get(uid types.UID) (*v1.Pod, bool) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	pod, ok := s.pods[uid]
+	if !ok {
+		return nil, false
+	}
+	return pod.DeepCopy(), true
+}
+
+// List returns all pods.
+func (s *PodsServer) List() []*v1.Pod {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	pods := make([]*v1.Pod, 0, len(s.pods))
+	for _, pod := range s.pods {
+		pods = append(pods, pod.DeepCopy())
+	}
+	sort.Slice(pods, func(i, j int) bool {
+		return pods[i].UID < pods[j].UID
+	})
+	return pods
+}
+
+// applyFieldMask uses Go's standard reflection to copy fields from src to dest based on the mask.
+// It supports traversing into slices to select specific fields from the elements.
+func ApplyFieldMask(mask *fieldmaskpb.FieldMask, src, dest interface{}) error {
+	srcVal := reflect.ValueOf(src)
+	destVal := reflect.ValueOf(dest)
+
+	if srcVal.Kind() != reflect.Ptr || destVal.Kind() != reflect.Ptr {
+		return fmt.Errorf("src and dest must be pointers")
+	}
+
+	for _, path := range mask.GetPaths() {
+		parts := strings.Split(path, ".")
+		if err := applySinglePath(srcVal, destVal, parts); err != nil {
+			// Note: This warning includes the full path, which is more helpful for debugging.
+			klog.Warningf("Failed to apply field mask path %q: %v", path, err)
+		}
+	}
+	return nil
+}
+
+// applySinglePath recursively traverses the src and dest objects according to the path parts,
+// creating nested objects in dest as needed and handling slice traversal.
+// `dest` must always be a pointer type.
+func applySinglePath(src, dest reflect.Value, path []string) error {
+	// Invariant: dest is always a pointer.
+	for i, part := range path {
+		// Source can be a pointer or a struct. If pointer, dereference it.
+		if src.Kind() == reflect.Ptr {
+			if src.IsNil() {
+				return nil // Source path is nil, nothing to copy.
+			}
+			src = src.Elem()
+		}
+
+		// Ensure dest is a non-nil pointer to a struct.
+		if dest.Kind() != reflect.Ptr {
+			return fmt.Errorf("internal error: dest is not a pointer at path part %q", part)
+		}
+		if dest.IsNil() {
+			dest.Set(reflect.New(dest.Type().Elem()))
+		}
+		destStruct := dest.Elem()
+
+		// We can only traverse structs.
+		if src.Kind() != reflect.Struct {
+			return fmt.Errorf("source is not a struct at path part %q", part)
+		}
+		if destStruct.Kind() != reflect.Struct {
+			return fmt.Errorf("destination is not a struct at path part %q", part)
+		}
+
+		// Find the field in the source struct.
+		// First, try to match by JSON tag. If that fails, fall back to Go field name.
```

Comment:
fieldmasks become part of the API surface ... I'm somewhat surprised to see an ad hoc implementation of it against REST API types here ... are we sure this works reliably across things like custom json marshaling (for metav1.Time, Quantity, etc) and inlined fields?

Reply:
We have unit tests for Time.  Which inlined fields were you thinking of?  Only mention of inlining I see in the Pod spec is ephemeral containers, which I added a unit test for.